### PR TITLE
Very basic support of RFC 5958

### PR DESCRIPTION
**Pull request by beldmit**
RFC 5958 extends RFC 5208 allowing a bit string incapsulating public key to be
present in PKCS#8 bundle. It is optional and implicit.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by kroeckx** [comment]
Should we rename it to OneAsymmetricKey, instead of the PKCS8_PRIV_KEY_INFO? Or have both?

It says that in that case the version must be 2. Should we enforce that somewhere?

I would also like to see an example of this being used somewhere.

**Comment by beldmit** [comment]
> Should we rename it to OneAsymmetricKey, instead of the PKCS8_PRIV_KEY_INFO? Or have both?

I don't know. I don't think that renaming is a good idea.

> It says that in that case the version must be 2. Should we enforce that somewhere?

Not now, I think.
 
> I would also like to see an example of this being used somewhere.

I have such an example for GOST keys. If I didn't have such an example, I'd like to avoid the modification the code.

**Comment by levitte** [comment]
> Should we rename it to OneAsymmetricKey, instead of the PKCS8_PRIV_KEY_INFO? Or have both?

Good question.  That has effects on a number of functions, but is possible if we retain compatibility macros.
Also, to be noted from RFC 5958:

```
PrivateKeyInfo ::= OneAsymmetricKey
```

**Comment by levitte** [comment]
> It says that in that case the version must be 2. Should we enforce that somewhere?

I think we should, and it shouldn't even be hard.  Something like this, perhaps?

``` diff
diff --git a/crypto/asn1/p8_pkey.c b/crypto/asn1/p8_pkey.c
index dee188519c..9da06c4743 100644
--- a/crypto/asn1/p8_pkey.c
+++ b/crypto/asn1/p8_pkey.c
@@ -23,6 +23,15 @@ static int pkey_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
         if (key->pkey)
             OPENSSL_cleanse(key->pkey->data, key->pkey->length);
     }
+    if (operation == ASN1_OP_D2I_POST) {
+        PKCS8_PRIV_KEY_INFO *key = (PKCS8_PRIV_KEY_INFO *)*pval;
+
+        if (key->version == 0 && key->pubkey != NULL) {
+            ERR_raise_data(ERR_LIB_ASN1, ERR_R_UNSUPPORTED,
+                           "Version 1 PKCS#8 doesn't support public key");
+            return 0;
+        }
+    }
     return 1;
 }
 
```

**Comment by beldmit** [comment]
> > Should we rename it to OneAsymmetricKey, instead of the PKCS8_PRIV_KEY_INFO? Or have both?
> 
> Good question. That has effects on a number of functions, but is possible if we retain compatibility macros.
> Also, to be noted from RFC 5958:
> 
> ```
> PrivateKeyInfo ::= OneAsymmetricKey
> ```

Personally I prefer to add a comment to a PKCS8_PRIV_KEY_INFO definition to make it greppable

**Comment by levitte** [comment]
> Personally I prefer to add a comment to a PKCS8_PRIV_KEY_INFO definition to make it greppable

Fine by me.  You're the coder, you get to decide :wink:

**Comment by beldmit** [comment]
Is it worth providing a getter/setter for  the `pubkey` field?

**Comment by mo-ang** [comment]
> 
> 
> Is it worth providing a getter/setter for the `pubkey` field?

I think a getter/setter could be very useful, similar to the `PKCS8_pkey_*()` functions.
A setter function would probably have to implicitly set version 2.

Additional thought:
When converting an `EVP_PKEY` object to `PKCS8_PRIV_KEY_INFO` by calling the `EVP_PKEY2PKCS8()` function, only the private key field is be considered and version 1 is passed explicitly to the `PKCS8_pkey_set0()` function. (see e.g. in `rsa_priv_encode()`)

Is it also worth to support converting `EVP_PKEY` key pairs to an RFC5958-conform (version 2) PKCS8 structure?

To support that I see two choices:
1. Adapt all the `evp_pkey_asn1_method_st::priv_encode` functions to set version 2 and to consider the new public key field - only if a public key is present in EVP_PKEY. Following this, `EVP_PKEY2PKCS8()` would create RFC5958 structures for key pairs per default.
2. Introduce a new function as alternative to `EVP_PKEY2PKCS8()`.

Am I missing a third way?

**Comment by davidben** [comment]
>> It says that in that case the version must be 2. Should we enforce that somewhere?
>
> Not now, I think.

Yes, this should be enforced and enforced from the beginning. This is new code, so this is your one chance to get the right behavior in without compatibility risks. OpenSSL should correctly enforce the version on both parse and serialization.

In terms of making new keys, changing the behavior of `EVP_PKEY2PKCS8` would not be a backwards-compatible change since it would no longer interoperate with v1-only parsers, like OpenSSL itself.

**Comment by csim-20** [comment]
> Should we rename it to OneAsymmetricKey, instead of the PKCS8_PRIV_KEY_INFO? Or have both?
> 
> It says that in that case the version must be 2. Should we enforce that somewhere?
> 
> I would also like to see an example of this being used somewhere.

**Comment by beldmit** [comment]
Set beta1 milestone to get some evaluation if we really want this feature

**Comment by mattcaswell** [comment]
OTC: Moving this to Post 3.0 milestone

**Comment by beldmit** [comment]
I'm OK to revive this PR (it even should be mergeable as is)

**Comment by t8m** [comment]
IMO this needs a test and documentation

**Comment by beldmit** [comment]
@t8m I am not sure about both. 

I can document in in NEWS/CHANGES, but I don't know where is a relevant place to put it into the man pages.

And also we need some test encoder/decoder to test it, right?

**Comment by beldmit** [comment]
In my GOST code I just ignored the pubkey part so his code was generally for parsing-only purpose. The decision can be done on the provider level. We can 
 * validate the key
   * throw an error in case of mismatch
   * silently recalculate it when possible
 * ignore it

**Comment by levitte** [comment]
Considering that this pretty much undocumented so far, congratulations, you get to add new documentation!

I'd suggest starting with a manual file that's named like the type.  With what we have now, that would be `doc/man3/PKCS8_PRIV_KEY_INFO.pod`.

Side note: to be frank, I'd rather have a type name without PKCS8 in it, 'cause RFC 5958 isn't PKCS#8, and having the name of the (now, not so) current standard wasn't a great idea...  but yeah, I get it, backward compat...

**Comment by levitte** [comment]
Re how to use this, I wonder if this could be left to the encoders?  For RSA, DH, DSA and EC, there's really no need to save the public key (either because the private key bits include the public key bits, or because the public key can be generated from the private key), so they could encode using version 1 (good old RFC 5208), while I assume that for example the OQS encoders would encode using version 2 (RFC 5958).  As I understand it, that would be desirable, no?

Do note that for the encoders that output PrivateKeyInfo file, that's exactly the structure name we use, without qualifying what version they use.

**Comment by beldmit** [comment]
OK. I will try to make some documentation for it.

**Comment by beldmit** [comment]
Looks like encoder/decoder will also need get0/set0 functions for public key.

@baentsch will it be enough for your purpose?

**Comment by levitte** [comment]
> Looks like encoder/decoder will also need get0/set0 functions for public key.

Er, why?  Shouldn't that end up in the `EVP_PKEY` they pull data from or put data in?

**Comment by beldmit** [comment]
I'm not sure, will look tomorrow.

The decision if the public key should be put in the structure should be done on the encoder level, but the public key can appear in the EVP_PKEY object even if we don't want to save it...

**Comment by levitte** [comment]
Oh.  Thing is that the encoders themselves handle the internal type, i.e. what the providers has intimate knowledge of.  If you look at this section, you can see the `EC_KEY` is used.  I totally expect the key-to-something encoder to have that sort of knowledge of the keys it encodes, i.e. the encoders for, say, the OQS stuff will inherently have to be part of the OQS provider.

Exactly how the public key is made accessible to the OQS encoder is 100% up to the OQS provider and no one else.

**Comment by beldmit** [comment]
@levitte sorry I still don't get.

I presume we have an EVP_PKEY object having both private and public part

https://github.com/openssl/openssl/blob/080bd08fd32608b4f2edfa4b1e87e199b08a8835/crypto/evp/evp_pkey.c#L134-L150 is the code converting it to PKCS* object

If I understand correctly, it invokes

https://github.com/openssl/openssl/blob/080bd08fd32608b4f2edfa4b1e87e199b08a8835/providers/implementations/encode_decode/encode_key2any.c#L80-L101

And we also want to set a public key part - and the current code seems not capable for it.

We can probably manage the exportability via passing the selection to `OSSL_ENCODER_CTX_new_for_pkey` which is currently set to `OSSL_KEYMGMT_SELECT_ALL`

**Comment by levitte** [comment]
Ok, first thing first: *the OpenSSL encoders don't need to be changed*, for the simple reason that there's no need to include the public key in a PKCS#8 blob for the key types that OpenSSL providers currently support.

The OQS provider is a separate project, so exactly how the PKCS#8 blob is generated by their encoder is entirely up to them.

That being said, we should add support for adding the public key in some of the functions in `crypto/asn1/p8_pkey.c`.  For example, I could see these added functions:

``` C
int PKCS8_pkey_set0_ex(PKCS8_PRIV_KEY_INFO *priv, ASN1_OBJECT *aobj,
                       int version, int ptype, void *pval,
                       unsigned char *privenc, int privenclen,
                       unsigned char *pubenc, int pubenclen);
int PKCS8_pkey_get0_ex(const ASN1_OBJECT **ppkalg,
                       const unsigned char **pprivk, int *pprivklen,
                       const unsigned char **ppubk, int *ppubklen,
                       const X509_ALGOR **pa, const PKCS8_PRIV_KEY_INFO *p8);
```

That way, an encoder that supports or needs to set the public key as well can make a separate DER encoding of it in whatever manner it wants, and populate the v2 PKCS8 structure with the help of `PKCS8_pkey_set0_ex()`

**Comment by beldmit** [comment]
I don't think we should change OpenSSL encoders, the question is about extra handles. 

Let me implement it the way you propose.

File: crypto/asn1/p8_pkey.c [hunk 1]
```diff
@@ -23,48 +23,85 @@ static int pkey_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
...
-    if (pk) {
-        *pk = ASN1_STRING_get0_data(p8->pkey);
-        *ppklen = ASN1_STRING_length(p8->pkey);
+    if (pprivk) {
+        *pprivk = ASN1_STRING_get0_data(p8->pkey);
+        *pprivklen = ASN1_STRING_length(p8->pkey);
+    }
+    if (ppubk) {
+        *ppubk = ASN1_STRING_get0_data(p8->pubkey);
+        *ppubklen = ASN1_STRING_length(p8->pubkey);
```
Comment (levitte) [inline crypto/asn1/p8_pkey.c:83]: This should depend on `p8->version > 0`.  When `p8->version == 0`, those fields should be cleared.  So, something like this perhaps?

```suggestion
    if (ppubk) {
        if (p8->version > 0) {
            *ppubk = ASN1_STRING_get0_data(p8->pubkey);
            *ppubklen = ASN1_STRING_length(p8->pubkey);
        } else {
            *ppubk = NULL;
            *ppubklen = 0;
        }
```

Comment (beldmit) [reply crypto/asn1/p8_pkey.c]: Thanks! Pushed

**Comment by beldmit** [comment]
The existing tests passed. Do we need any new?

**Comment by baentsch** [comment]
> @baentsch will it be enough for your purpose?

Well, [the original ask](https://github.com/open-quantum-safe/oqs-provider/issues/194#issuecomment-1724878063) was not specifically for inclusion of a public key in PKCS#8 but just "some" output of the public key. In sum I think this feature is great but can't completely replace/solve #22132 -- unless there's a simple command to extract the public key from a PKCS#8 file (?)

**Comment by levitte** [comment]
> ... unless there's a simple command to extract the public key from a PKCS#8 file (?)

Add the option `-pubout` to `openssl pkcs8`?

**Comment by beldmit** [comment]
@baentsch could you make a PKCS#8 file containing a public key using this API so I could add this option?

**Comment by baentsch** [comment]
> @baentsch could you make a PKCS#8 file containing a public key using this API so I could add this option?

Will try to do this. Stay tuned.

**Comment by baentsch** [comment]
> Will try to do this. Stay tuned.

@beldmit You have reviewed #22173 and thus may foresee the following suggestion borne of my inherent laziness (to try to avoid learning APIs I'll never use again): Would you mind I add equivalent code to this PR too instead of creating some sample PKCS#8 file via P8 API? That would also give us the ability to cross-check (aka test) the results of both features (via command line option `-outpubkey` in `genpkey` and `pkcs8`): This should then allow us to exercise the P8 API automatically (without me having to learn it :), right? Added benefit: You don't have to add the option any more :)

**Comment by beldmit** [comment]
I am lazy and so don't mind :)

The question is - should we just print out the public key if it is present or do anything else

**Comment by baentsch** [comment]
> should we just print out the public key if it is present or do anything else

Continuing on the "laziness thought train" I wouldn't print it or do anything else: Why confuse people with information they're not explicitly asking for? Those that do (know what they do :) will pass `-outpubkey` and get what they want.

**Comment by beldmit** [comment]
Sorry for unclarity. I didn't mean to print, I meant output on request. But the question is basically the same.

We have a PKCS#8 file. I see 3 situations:
1. There is no public key in the file.
2. There is a correct public key in the file
3. There  is an incorrect key in the file.

Do we try calculating the public key in (1)?
Do we try to distinguish (2) and (3)?

**Comment by levitte** [comment]
Who is "we"? It's really up to the provider to decide if the public key needs to be present or not, possibly with a hint from the user. It's definitely not a decision libcrypto should make.

**Comment by beldmit** [comment]
It's up to the provider to include the public key part into the OneAsymmetricKey bundle. I wonder what is a proper behavior when we call `openssl pkcs8 -outpubkey`

**Comment by levitte** [comment]
In my mind, `openssl pkcs8 -pubout -in new key.p8` would output the public key in the form the user asked for. Not sure how `-outpubkey` came to be in this discussion...

But quite frankly, I think we may have to have a separate OQS discussion, 'cause there's something about the public key in that general idea that doesn't quite fit.

**Comment by levitte** [comment]
@baentsch, re being "lazy", what you will need to do is to modify `key_to_p8info`:

https://github.com/open-quantum-safe/oqs-provider/blob/9d4faab2d330df8b032995ed58f757c25b1c68c7/oqsprov/oqs_encode_key2any.c#L77-L100

All it would need, I think, is an extra argument `i2d_of_void *pubk2d`, which would be a function to get the DER encoding of the *public* part of `key`.  Use that with a second der buffer (you could call it `pubder`), and then call this instead of your current `PKCS8_pkey_set0` call.

``` C
        || PKCS8_pkey_set0_ex(p8info, OBJ_nid2obj(key_nid), 0,
                              V_ASN1_UNDEF, NULL, der, derlen,
                              pubder, pubderlen)
```

**Comment by nhorman** [comment]
given that this PR needs a rebase, and has not been updated in almost a year, I'm closing for inactivity.  Please reopen if you would like to pick this back up

**Comment by levitte** [comment]
I'd like to see this picked back up.  @beldmit, have you given up on this?  In that case, I can see myself taking this over (new PR in the case, make from my copy of this branch)

**Comment by beldmit** [comment]
Do we have real-life examples for this code? My interest was initially motivated by some GOST examples I used to maintain, but now I have nothing to do with it

**Pull request closed without merging**